lib/src/git_subprocess.rs

```diff
@@ -0,0 +1,484 @@
+use std::ffi::OsStr;
+use std::fmt;
+use std::num::NonZeroU32;
+use std::path::Path;
+use std::path::PathBuf;
+use std::process::Command;
+use std::process::Output;
+use std::process::Stdio;
+
+use thiserror::Error;
+
+use crate::git::GitFetchError;
+use crate::git::GitPushError;
+
+/// New-type for comparable io errors
+pub struct IoError(std::io::Error);
+impl PartialEq for IoError {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.kind().eq(&other.0.kind())
+    }
+}
+impl std::error::Error for IoError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        Some(&self.0)
+    }
+}
+impl std::fmt::Display for IoError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+impl std::fmt::Debug for IoError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+#[derive(Error, Debug, PartialEq)]
+pub enum GitError {
+    #[error("No git remote named '{0}'")]
+    NoSuchRemote(String),
+    #[error("Could not find `git` executable: provided path {0:?}")]
+    GitNotFound(Option<String>),
+    #[error("Failed to fork git process")]
+    GitForkError(#[from] IoError),
+    #[error("Git process failed: {0}")]
+    GitExternalError(String),
+    #[error("Failed to convert path {0:?} to a string")]
+    PathConversionError(std::path::PathBuf),
+}
+impl From<std::io::Error> for GitError {
+    fn from(value: std::io::Error) -> Self {
+        GitError::GitForkError(IoError(value))
+    }
+}
+
+impl From<GitError> for GitFetchError {
+    fn from(value: GitError) -> Self {
+        match value {
+            GitError::NoSuchRemote(remote) => GitFetchError::NoSuchRemote(remote),
+            GitError::GitNotFound(git_path) => GitFetchError::GitNotFound(git_path),
+            GitError::GitForkError(io_err) => GitFetchError::GitForkError(io_err.0),
+            GitError::GitExternalError(error) => GitFetchError::GitExternalError(error),
+            GitError::PathConversionError(path) => GitFetchError::PathConversionError(path),
+        }
+    }
+}
+
+impl From<GitError> for GitPushError {
+    fn from(value: GitError) -> Self {
+        match value {
+            GitError::NoSuchRemote(remote) => GitPushError::NoSuchRemote(remote),
+            GitError::GitNotFound(git_path) => GitPushError::GitNotFound(git_path),
+            GitError::GitForkError(io_err) => GitPushError::GitForkError(io_err),
+            GitError::GitExternalError(error) => GitPushError::GitExternalError(error),
+            GitError::PathConversionError(path) => GitPushError::PathConversionError(path),
+        }
+    }
+}
+
+/// Context for creating git subprocesses
+pub(crate) struct GitSubprocessContext<'a> {
+    git_dir: String,
+    work_tree: String,
+    git_path: &'a str,
+}
+
+impl<'a> GitSubprocessContext<'a> {
+    pub(crate) fn new(git_dir: impl ToString, work_tree: impl ToString, git_path: &'a str) -> Self {
```

Comment:
nit: `Into<String>` is safer. Various types support `ToString` for displaying purpose.